File: utilities/src/main/java/xerus/util/helpers/Row.java
package xerus.util.helpers;

import java.util.AbstractList;
import java.util.Arrays;

public class Row extends AbstractList<String> {
	
	private String[] data;
	private int size;
	
	public Row(int size, String... data) {
		this.data = data;
		this.size = size;
	}
	
	public Row(String... data) {
		this(data.length, data);
	}
	
	public String get(int col) {
		if (col >= data.length)
			return "";
		return data[col];
	}
	
	public String set(int col, String o) {
		String old = null;
		if (data.length < col)
			data = Arrays.copyOf(data, col + 1);
		else
			old = data[col];
		data[col] = o;
		return old;
	}
	
	@Override
	public int size() {
		return size;
	}
	
	@Override
	public String toString() {
		return "Row" + Arrays.toString(data);
	}
}

File: kotlin/src/xerus/ktutil/helpers/Row.kt
package xerus.ktutil.helpers

import java.util.*

class Row(override var size: Int, vararg data: String) : AbstractList<String>(), RandomAccess {
	constructor(vararg data: String) : this(data.size, *data)
	
	@Suppress("Unchecked_cast")
	private var data: Array<String> = data as Array<String>
	
	override fun get(index: Int): String =
			if (index >= data.size) "" else data[index]
	
	override fun set(index: Int, element: String): String? {
		var old: String? = null
		if (data.size < index) {
			data = Arrays.copyOf(data, index + 2).map { it ?: "" }.toTypedArray()
			size = data.size.coerceAtLeast(size)
		} else
			old = data[index]
		data[index] = element
		return old
	}
	
	override fun toString() = "Row" + Arrays.toString(data)
	
}